File: Open/max_sub.cpp
#include <bits/stdc++.h>
using namespace std;

/*
  Brian Grenier
  1545276
  bgrenier

  List any resources you used below (eg. urls, name of the algorithm from our code archive).
  Remember, you are permitted to get help with general concepts about algorithms
  and problem solving, but you are not permitted to hunt down solutions to
  these particular problems!

  N/A

  List any classmate you discussed the problem with. Remember, you can only
  have high-level verbal discussions. No code should be shared, developed,
  or even looked at in these chats. No formulas or pseudocode should be
  written or shared in these chats.

  N/A

  By submitting this code, you are agreeing that you have solved in accordance
  with the collaboration policy in CMPUT 403.
*/


using p = pair<int, int>;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, en;
    cin >> n;
    p arr[n+1];
    int last = -1;
    int lastind = 0;
    int temp;
    for (int i = 0; i < n; ++i) {
        cin >> temp;
        if (temp == last) {
            arr[lastind].second += 1;
        } else {
            arr[lastind] = {temp, 1};
            ++lastind;
        }
    }
    arr[lastind+1] = {-1, 1};

    int start, m, t;
    cin >> en;
    unordered_set<int> s;

    for (int i = 0; i < en; ++i) {
        s.clear();
        cin >> start >> m;
        --start;

        if (m == n) {
            cout << n - start << endl;
            for (int k = 0; k < m; ++k) {
                cin >> start;
            }
            continue;
        }
        
        s.reserve(m);
        for (int k = 0; k < m; ++k) {
            cin >> t;
            s.insert(t);
        }

        int count = 0;
        int j = 0;

        while (start != 0) {
            if (arr[j].second <= start) {
                start -= arr[j].second;
                ++j;
            } else {
                if (s.count(arr[j].first) > 0) {
                    count += arr[j].second - start;
                    ++j;
                    break;
                } else {
                    goto nothing;
                }
            }
        }

        for (; j <= lastind; ++j) {
            
            if (s.count(arr[j].first) > 0) {
                count += arr[j].second;
            } else {
                break;
            }

        }
        nothing:

        
        cout << count << endl;
    }
}

File: Open/max_sub.py
import sys
"""
  Brian Grenier
  1545276
  bgrenier

  List any resources you used below (eg. urls, name of the algorithm from our code archive).
  Remember, you are permitted to get help with general concepts about algorithms
  and problem solving, but you are not permitted to hunt down solutions to
  these particular problems!

https://stackoverflow.com/questions/9501337/binary-search-algorithm-in-python

  List any classmate you discussed the problem with. Remember, you can only
  have high-level verbal discussions. No code should be shared, developed,
  or even looked at in these chats. No formulas or pseudocode should be
  written or shared in these chats.

  N/A

  By submitting this code, you are agreeing that you have solved in accordance
  with 
"""

def binary_search(array, target):
    lower = 0
    upper = len(array)
    while lower < upper:   # use < instead of <=
        x = lower + (upper - lower) // 2
        val = array[x]
        if target == val:
            return x
        elif target > val:
            if lower == x:   # these two are the actual lines
                break        # you're looking for
            lower = x
        elif target < val:
            upper = x

if __name__ == '__main__':
    len_seq = int(sys.stdin.readline())
    nums = list(map(int, sys.stdin.readline().split()))
    num_q = int(sys.stdin.readline())

    for i in range(num_q):
        q = list(map(int, sys.stdin.readline().split()))
        index = q[0]
        query = q[2:]
        query.sort()
        count = 0
        for k in range(index-1, len(nums)):
            if not binary_search(query, nums[k]) is None:
                count += 1
            else:
                break
        print(count)